Add AI Code Assistant Integration with Gemini API

hyc17003 opened the pull request:
## Summary

This PR introduces an AI-powered code assistant to the PeerPrep collaboration platform, providing real-time programming help during coding sessions. The assistant uses Google's Gemini 2.5 Flash model to offer code explanations, debugging assistance, optimization suggestions, and algorithm guidance.

**Key Features:**
- 💬 Interactive chat interface with markdown-formatted responses
- 🔄 Language-aware assistance based on current editor language
- 📝 Automatic code context from editor
- 🎨 Toggle between peer chat and AI assistant
- ⚡ Fast responses using Gemini 2.5 Flash (free tier)

## Changes Overview

### Backend
- **New AI Service** ([ai-service/](ai-service/))
  - Express.js microservice for handling AI requests
  - Google Gemini API integration (`gemini-2.5-flash`)
  - Conversation history support (last 5 messages as context)
  - Safety guardrails in system prompt (rejects non-programming requests)

### Frontend
- **New Components**
  - [`AiAssistPanel.tsx`](frontend/src/app/components/collab/AiAssistPanel.tsx) - Main AI chat interface with markdown rendering
  - UI features: message history, loading states, copy-to-clipboard
  
- **Modified Components**
  - [`collab/page.tsx`](frontend/src/app/collab/page.tsx) - Added toggle between chat and AI assist
  - [`CodingComponentWrapper.tsx`](frontend/src/app/components/collab/CodingComponentWrapper.tsx) - Enhanced to expose editor instance and language state
  - [`CodingComponent.tsx`](frontend/src/app/components/collab/CodingComponent.tsx) - Passes editor reference and language to parent

### Infrastructure
- Added AI service to [`docker-compose.yml`](docker-compose.yml) on port 8086
- New service API client in [`frontend/src/services/aiServiceApi.ts`](frontend/src/services/aiServiceApi.ts)

## Why Collab Page Was Modified

The collab page and coding components were restructured to enable **language-specific AI assistance**. The wrapper pattern allows:

1. **Language context propagation**: The editor's current language (JavaScript, Python, Java, etc.) is lifted up from the Monaco editor instance to the collab page
2. **Editor instance access**: AI assistant needs direct access to the code editor to retrieve current code content
3. **State coordination**: Both the language selector and AI panel need to share the same language state for accurate assistance

This architecture ensures the AI assistant receives proper context about what programming language the user is working with, enabling more accurate and relevant suggestions.

## Known Limitations & Future Work

### ⚠️ Current Limitations

1. **No Conversation Persistence**
   - Chat history is cleared on page refresh/navigation
   - Users lose context when switching between chat/AI assist
   - **Future**: Consider adding backend storage (Redis/MongoDB) for session persistence (if we deem it necessary)

2. **Scroll Container Issue**
   - Messages may scroll beyond the AI assist panel boundaries
   - **TODO**: Fix overflow/height constraints to contain scrolling within the panel

3. **Prompt Engineering Needs Refinement**
   - System prompts need testing with various coding scenarios
   - **TODO**: Iterate on prompt templates based on user feedback and edge cases
   - Consider adding few-shot examples for better response quality

## Technical Details

**Model**: Google Gemini 2.5 Flash
- Free tier: 60 requests/min, 1,500 requests/day
- Fast response times (~1-2 seconds)
- Context window: Includes current code + last 5 messages

## Testing

**Manual Testing**:
- ✅ Toggle between peer chat and AI assist
- ✅ Send questions with/without code in editor
- ✅ Language switching reflects in AI responses
- ✅ Markdown rendering (code blocks, inline code, lists)
- ✅ Copy-to-clipboard functionality

**Setup Required:**:
1. Add `ai-service/.env` file, which contains the Gemini Api Key
2. Build and start AI service on docker
4. Navigate to collab page and click "AI Assist"

## Screenshots

![ai chat](https://github.com/user-attachments/assets/98c6a09d-9b6e-41c6-a261-ec692ac0386a)

![ai chat 2](https://github.com/user-attachments/assets/e2dc71e2-8c99-4663-b866-a7a0c3f30d23)

FooChao commented on frontend/Dockerfile:
```diff
@@ -24,6 +24,14 @@ COPY . .
+ARG NEXT_PUBLIC_USER_SERVICE_URL
+
+# Set them as environment variables during build
+ENV NEXT_PUBLIC_API_GATEWAY_BASE_URL=$NEXT_PUBLIC_API_GATEWAY_BASE_URL
```
can i check what is the purpose of this when u dun pass i the args for it?
The reason why it works after u add this (when not defined on Dockerfile) is because u nulled it and then it bypassed back to http://user-service:4000 
<img width="741" height="162" alt="image" src="https://github.com/user-attachments/assets/e3f6986f-5642-4feb-8950-24e34d940185" />
the main reason i using user-service:4000 is because next.js is stupid when used with docker so i have to find a way to bypass it
<img width="1370" height="313" alt="image" src="https://github.com/user-attachments/assets/f4c61dd1-8bdc-4975-b4af-30b353db1e97" />

so if u have no particular reason to do it can u change the link in .env.local back to http://user-service:4000 
I have set it such that only when it is middleware it calls this all the rest is via api gateway. (on deployment is everything via api gateway)

FooChao commented on frontend/src/services/aiServiceApi.ts:
```diff
@@ -0,0 +1,72 @@
+ */
+
+const AI_SERVICE_URL =
+  process.env.NEXT_PUBLIC_AI_SERVICE_URL || "http://localhost:8086";
```
can i check why u not using the api gateway since it is up already?

hyc17003 commented:
> LGTM! Found a small bug but maybe not super high priority to fix (i think??) when a code is provided a copy button appeared but when u press the copy it copy not the code but entire message.

fixed! added the language within the prompt as well so the user can get language-specific help without having typed any code
